File: testing/c/floattest.c
// floatdemo.c — IEEE 754 soft-float demo for the MPU
// All four arithmetic operations, int↔float conversion, and comparison.

void main() {
    float pi = 3.14159;
    float r = 5.0;
    float area = pi * r * r;
    printf("area of circle r=5: %d (expect 78)\n", ftoi(area));

    float c = 100.0;
    float f = c * 1.8 + itof(32);
    printf("100 C in F: %d (expect 212)\n", ftoi(f));

    float a = 355.0;
    float b = 113.0;
    float approx_pi = a / b;
    printf("355/113 = %d.%d (expect 3.14)\n",
        ftoi(approx_pi),
        ftoi((approx_pi - itof(ftoi(approx_pi))) * 100.0));

    // ADC voltage conversion
    int raw = adc_read();
    float voltage = itof(raw) * 3.3 / 4095.0;
    printf("ADC raw=%d voltage*100=%d mV\n", raw,
        ftoi(voltage * 100.0));

    // Comparison
    if (fcmp(pi, 3.0) > 0) printf("pi > 3: yes\n");
    if (fcmp(2.71, pi) < 0) printf("e < pi: yes\n");
}

File: testing/floattest.pas
program floattest;

var
  x, y, z : real;
  n : integer;

function area(r: real): real;
begin
  area := 3.14159 * r * r
end;

begin
  writeln('--- literals and arithmetic ---');
  x := 3.14;
  y := 2.0;
  writeln('x = ', x);
  writeln('y = ', y);
  writeln('x + y = ', x + y);
  writeln('x - y = ', x - y);
  writeln('x * y = ', x * y);
  writeln('x / y = ', x / y);

  writeln('--- unary and comparison ---');
  z := -1.5;
  writeln('-1.5 = ', z);
  if z < 0.0 then writeln('z < 0: yes');
  if x > y then writeln('x > y: yes');
  if x = 3.14 then writeln('x = 3.14: yes');

  writeln('--- mixed int/real promotes ---');
  n := 5;
  x := itof(n) + 0.5;
  writeln('itof(5) + 0.5 = ', x);

  writeln('--- trunc and sqrt ---');
  writeln('trunc(3.7) = ', trunc(3.7));
  writeln('sqrt(2.0) = ', sqrt(2.0));
  writeln('sqrt(144.0) = ', sqrt(144.0));

  writeln('--- area of circle r=5 ---');
  writeln('area = ', area(5.0))
end.
